### PR TITLE
Niad 2858: storing negative ack error code in DB

### PR DESCRIPTION
**Pull request by ole4ryb**
## What

Capturing negative ack error code in DB

## Why

PS Adaptor doesn't record/store original error code sent by an incumbent in case of a failure. There is a requirement by NHS England to send failure details hence there should be some mechanism to capture it in PS Adaptor in order to give an access to such information when required.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Internal change (non-breaking change with no effect on the functionality affecting end users)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have updated the [Changelog](/CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users
- [ ] A corresponding change has been made to the [Mapping Documentation repository][mapping-docs]

[mapping-docs]: https://github.com/NHSDigital/patient-switching-adaptors-mapping-documentation

### CODE REVIEW

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/scheduled/EHRTimeoutHandler.java [hunk 1]
```diff
@@ -171,7 +171,7 @@ private void handleMigrationTimeout(PatientMigrationRequest migrationRequest, NA
             LOGGER.error("Error retrieving persist duration: [{}]", e.getMessage());
         } catch (JsonProcessingException | SAXException | DateTimeParseException | JAXBException e) {
             LOGGER.error("Error parsing inbound message from database");
-            migrationStatusLogService.addMigrationStatusLog(EHR_GENERAL_PROCESSING_ERROR, conversationId, null);
+            migrationStatusLogService.addMigrationStatusLog(EHR_GENERAL_PROCESSING_ERROR, conversationId, null, "99");
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/scheduled/EHRTimeoutHandler.java:174]: What was the thinking behind "99" in this case. I'm not sure I have an alternative suggestion as to what we could do here, just curious as to the thinking.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/SendEhrExtractRequestHandler.java [hunk 2]
```diff
@@ -43,12 +43,12 @@ public boolean prepareAndSendRequest(TransferRequestMessage message) {
             LOGGER.debug(response);
         } catch (WebClientResponseException wcre) {
             LOGGER.error("Received an ERROR response from MHS: [{}]", wcre.getMessage());
-            migrationStatusLogService.addMigrationStatusLog(MigrationStatus.EHR_EXTRACT_REQUEST_ERROR, conversationId, null);
+            migrationStatusLogService.addMigrationStatusLog(MigrationStatus.EHR_EXTRACT_REQUEST_ERROR, conversationId, null, "2");
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/SendEhrExtractRequestHandler.java:46]: "2" is mysterious. I thought GP2GP errors were all 2 digits.

Also, I wonder if it makes sense to refererence an `enum` of valid GP2GP error codes. That way this line of code would become more understandable.

Update: Looking at the NACKReason enum, I can't see a 2. 🤔

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/SendContinueRequestHandler.java [hunk 3]
```diff
@@ -41,7 +41,7 @@ public void prepareAndSendRequest(ContinueRequestData data) {
             mhsClientService.send(request);
         } catch (WebClientResponseException webClientResponseException) {
             LOGGER.error("Received an ERROR response from MHS: [{}]", webClientResponseException.getMessage());
-            migrationStatusLogService.addMigrationStatusLog(MigrationStatus.CONTINUE_REQUEST_ERROR, data.getConversationId(), null);
+            migrationStatusLogService.addMigrationStatusLog(MigrationStatus.CONTINUE_REQUEST_ERROR, data.getConversationId(), null, "8");
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/SendContinueRequestHandler.java:44]: Similar comment to the "2" comment below.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java [hunk 4]
```diff
@@ -1280,7 +1280,7 @@ public void When_HandleMessage_With_AttachmentProcessingErrorNotStorageCause_Exp
             assertThat(nackMessageDataCaptor.getValue().getNackCode()).isEqualTo(LARGE_MESSAGE_ATTACHMENTS_NOT_RECEIVED.getCode());
 
             verify(migrationStatusLogService, times(1))
-                .addMigrationStatusLog(LARGE_MESSAGE_ATTACHMENTS_NOT_RECEIVED.getMigrationStatus(), CONVERSATION_ID, null);
+                .addMigrationStatusLog(LARGE_MESSAGE_ATTACHMENTS_NOT_RECEIVED.getMigrationStatus(), CONVERSATION_ID, null, "31");
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandlerTest.java:1283]: What's interesting here is that `LARGE_MESSAGE_ATTACHMENTS_NOT_RECEIVED` already has the code associated with it.

Perhaps this suggests we should have a new method called `.addNackMigrationStatusLog` which takes in a `NACKReason`. Thoughts?

File: db-connector/changelog/migration/14-add-error-code-column-to-migration_status_log-table.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+        xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog
+	http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.1.xsd
+	http://www.liquibase.org/xml/ns/dbchangelog-ext
+	http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd
+    http://www.liquibase.org/xml/ns/pro
+	http://www.liquibase.org/xml/ns/pro/liquibase-pro-4.1.xsd">
+
+    <changeSet id="14" author="ole4ryb">
+        <addColumn schemaName="public" tableName="migration_status_log">
+            <column name="error_code" type="varchar(255)">
```
Comment (adrianclay) [inline db-connector/changelog/migration/14-add-error-code-column-to-migration_status_log-table.xml:15]: I'm wondering if we should be clear in the name that this is a `GP2GP Error Code`.

File: gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandler.java [hunk 6]
```diff
@@ -150,7 +150,7 @@ public void handleMessage(InboundMessage inboundMessage, String conversationId)
             } else {
                 failMigration(conversationId, LARGE_MESSAGE_ATTACHMENTS_NOT_RECEIVED);
                 migrationStatusLogService.addMigrationStatusLog(
-                    LARGE_MESSAGE_ATTACHMENTS_NOT_RECEIVED.getMigrationStatus(), conversationId, null);
+                    LARGE_MESSAGE_ATTACHMENTS_NOT_RECEIVED.getMigrationStatus(), conversationId, null, "31");
```
Comment (adrianclay) [inline gp2gp-translator/src/main/java/uk/nhs/adaptors/pss/translator/task/COPCMessageHandler.java:153]: Mysterious 31.

File: gpc-api-facade/src/test/java/uk/nhs/adaptors/pss/gpc/util/fhir/OperationOutcomeUtilsTest.java [hunk 7]
```diff
@@ -15,16 +14,17 @@ public class OperationOutcomeUtilsTest {
...
 
         assertThat(result.getMeta().getProfile().get(0).equals(URI_TYPE)).isTrue();
-        assertThat(result.getIssueFirstRep().getCode()).isEqualTo(NOTSUPPORTED);
-        assertThat(result.getIssueFirstRep().getSeverity()).isEqualTo(ERROR);
-        assertThat(result.getIssueFirstRep().getDetails()).isEqualTo(details);
-        assertThat(result.getIssueFirstRep().getDiagnostics()).isEqualTo(diagnostics);
+        assertEquals(NOTSUPPORTED, result.getIssueFirstRep().getCode());
+        assertEquals(ERROR, result.getIssueFirstRep().getSeverity());
+        assertEquals(details, result.getIssueFirstRep().getDetails());
+        assertEquals(diagnostics, result.getIssueFirstRep().getDiagnostics());
```
Comment (adrianclay) [inline gpc-api-facade/src/test/java/uk/nhs/adaptors/pss/gpc/util/fhir/OperationOutcomeUtilsTest.java:27]: While I love a good refactor, mixing functionality with refactors within the same PR can be a bit unclear when looking at the change whether something has intentionally changed or not.

Comment (ole4ryb) [reply gpc-api-facade/src/test/java/uk/nhs/adaptors/pss/gpc/util/fhir/OperationOutcomeUtilsTest.java:27]: I just thought that this refactoring is quite simple and straightforward (all changes in one place) hence decided to include it in here

**Review by adrianclay — APPROVED** [review]

**Pull request merged**